### PR TITLE
#4431: mnist gs_demo benchmark

### PR DESCRIPTION
**Pull request by jayasuryamaganuru**
This PR contains benchmarking of mnist gs_demo for MNIST Handwritten Digit Classification test dataset.
model is tested with 1000 datapoints extracted from the dataset and the accuracy is listed below:

Metric: Accuracy
Accuracy: 62.2

### CODE REVIEW

**Review by muthutt — CHANGES_REQUESTED** [review]
request to add randomization (in repeatable way) to sample the data set for inference

File: models/experimental/mnist/tests/test_perf_accuracy_mnist.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    disable_persistent_kernel_cache()
+    first_key = "first_iter"
+    second_key = "second_iter"
+    third_key = "third_iter"
+    cpu_key = "ref_key"
+    iteration = 0
+    # Data preprocessing/loading
+    transform = transforms.Compose([transforms.ToTensor(), transforms.Normalize((0.1307,), (0.3081,))])
+    test_dataset = datasets.MNIST(root="./data", train=False, transform=transform, download=True)
+    dataloader = DataLoader(test_dataset, batch_size=1)
```
[the remark this answers is not in the document]
Comment (jayasuryamaganuru) [reply models/experimental/mnist/tests/test_perf_accuracy_mnist.py]: committed the suggestion, 
If shuffle=True isn't specified in Dataloader , randomization wasn't happening, even if we specify the generator.
commit link - https://github.com/tenstorrent-metal/tt-metal/compare/main...jayasurya/mnist_gs_demo_benchmarking

**Comment by jayasuryamaganuru** [comment]
When randomness introduced to the Dataloader with seed set to 213, there's slight increase in accuracy.

metric: Accuracy
Accuracy: 63.9

**Comment by jayasuryamaganuru** [comment]
Thank you for noting the lower accuracy. I shared the same concern, 
as we're using the torch model's state_dict for weights instead of starting with random values.

I evaluated the accuracy of the torch model and found it matches the accuracy of the tt model, both around 60%. 
After reviewing a [comment](https://github.com/tenstorrent-metal/tt-metal/blob/1de7f848d5c27b830ff4e2a669819d5bb5e6321e/models/experimental/mnist/tt/mnist_model.py#L61), I've become convinced that the model might not have been trained optimally.


> Thanks - Randomness pattern in dataloader for validation testing is important; besides that it seems like very low accuracy for MNIST model which typically runs at 99%; are the weights random or not trained well ? Just curious.
> […](#)
> On Thu, Dec 28, 2023 at 5:46 AM jayasuryamaganuru ***@***.***> wrote: When randomness introduced to the Dataloader with seed set to 213, there's slight increase in accuracy. metric: Accuracy Accuracy: 63.9 — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you commented.Message ID: ***@***.***>

**Review by boris-drazic — APPROVED** [review]

**Review by muthutt — APPROVED** [review]
LGTM

**Comment by jayasuryamaganuru** [comment]
@muthutt, the MNIST model weights have been updated in the WEKA path,
with a torch model that was trained to achieve an accuracy of **98.24%** on MNIST test dataset.
Now the tt_model is yielding an accuracy of **80.0%.**, for 1000 iterations

**Comment by jayasuryamaganuru** [comment]
@muthutt Yes, that's correct. However, the metric used for evaluation is accuracy, not F1 score

**Review by farbabi — APPROVED** [review]
